Initial contribution of the DSC Alarm Binding

RSStephens opened the pull request:
Initial contribution of the DSC Alarm Binding as per issue #99.

Signed-off-by: Russell Stephens <rsstephens@gmail.com> (Github: @RSStephens)

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/ESH-INF/thing/envisalinkbridge.xml:
```diff
@@ -0,0 +1,54 @@
+    xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 org.eclipse.smarthome.thing-description.xsd">
+
+	<bridge-type id="envisalinkbridge">
+		<label>EyezOn Envisalink Bridge</label>
```
Just give the product name here, i.e. remove the "Bridge".

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/ESH-INF/thing/envisalinkbridge.xml:
```diff
@@ -0,0 +1,54 @@
+		<config-description>
+			<parameter name="ipAddress" type="text" required="true">
+				<context>network_address</context>
+				<label>Envisalink Bridge IP</label>
```
we are already in the context of the Envisalink, so just call it "IP address" or "Hostname".

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/ESH-INF/thing/envisalinkbridge.xml:
```diff
@@ -0,0 +1,54 @@
+            </parameter>
+
+			<parameter name="password" type="text" required="false">
+                <label>Password</label>
```
also add
```
<context>password</context>
```
here.

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/ESH-INF/thing/envisalinkbridge.xml:
```diff
@@ -0,0 +1,54 @@
+
+			<parameter name="pollPeriod" type="integer" required="false" min="1" max="15">
+                <label>Poll Period</label>
+                <description>The Poll Period.</description>
```
in seconds?

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/ESH-INF/thing/envisalinkbridge.xml:
```diff
@@ -0,0 +1,54 @@
+		<description>This bridge represents the EyezOn Envlisalink 3/2D Ethernet interface.</description>
+
+        <channels>
+            <channel id="bridge_connection" typeId="bridge_connection"/>
```
I would assume you do not need this channel. If it is only to know whether the bridge is reachable or not, this is done through the Thing status ONLINE/OFFLINE instead.

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/ESH-INF/thing/it100bridge.xml:
```diff
@@ -0,0 +1,42 @@
+    xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 org.eclipse.smarthome.thing-description.xsd">
+
+	<bridge-type id="it100bridge">	
+		<label>DSC IT-100 Bridge</label>
```
again, remove the "Bridge".

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/ESH-INF/thing/it100bridge.xml:
```diff
@@ -0,0 +1,42 @@
+		<description>This bridge represents the DSC IT-100 Serial interface.</description>
+
+        <channels>
+            <channel id="bridge_connection" typeId="bridge_connection"/>
```
remove the channel

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/ESH-INF/thing/keypad.xml:
```diff
@@ -0,0 +1,94 @@
+        <item-type>Number</item-type>
+        <label>Keypad Ready LED</label>
+        <description>Keypad Ready LED (0=Off, 1=On, 2=Flashing)</description>
+        <state min="0" max="2" step="1" readOnly="true"></state>
```
Instead of using min, max and step, you should rather directly list the options with `<options>` - you can then even provide labels for the numbers.

RSStephens replied:
I remember looking into using ```<options>``` for this, but the requirement was the item type had to be String.  Can ```<options>``` be used with an item type of Number?

kaikreuzer replied:
> Can `<options>` be used with an item type of Number?

I would expect it to. Everything else I would consider a bug.

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/ESH-INF/thing/keypad.xml:
```diff
@@ -0,0 +1,94 @@
+        <description>Represents the central administrative unit of the DSC Alarm System.</description>
+
+        <channels>
+            <channel id="keypad_ready_led" typeId="keypad_ready_led"/>
```
All your channels seem to be of the same type - an signalling led with is on/off/flashing.
Since a few weeks, you can add label and descriptions directly to the channel definitions here, so there is probably no need to define different channel types, but one should be enough.

Try something like
```
<channel id="keypad_ready_led" typeId="led">
  <label>Keypad Armed LED</label>
  <description>Keypad Armed LED (0=Off, 1=On, 2=Flashing)</description>
</channel>
```

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/ESH-INF/thing/partition.xml:
```diff
@@ -0,0 +1,86 @@
+
+			<parameter name="partitionNumber" type="integer" required="true" min="1" max="8">
+				<label>Partition Number</label>
+				<description>The Partition Number (1-8).</description>
```
consider listing options, so that UIs can render a selection box for it.

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/ESH-INF/thing/panel.xml:
```diff
@@ -0,0 +1,194 @@
+
+    </thing-type>
+
+    <channel-type id="panel_message">
```
Are any of the channel types something that you would consider to not be the "main functionality" and thus could be marked as "advanced" (which will make it appear in the UI only on user request)?

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/ESH-INF/thing/zone.xml:
```diff
@@ -0,0 +1,92 @@
+    </thing-type>
+
+    <channel-type id="zone_status">
+        <item-type>Contact</item-type>
```
Does OPEN/CLOSED also make sense for a motion sensor?

RSStephens replied:
On the DSC Alarm a zone's status is either open or closed regardless of the device (contact, motion sensor, smoke detector, etc.) connected.  So, if a motion detector, for example, is triggered the zone is going to be shown as open.

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,24 @@
+Bundle-ClassPath: .
+Import-Package: com.google.common.collect,
+ gnu.io,
+ org.apache.commons.io;version="2.2.0",
```
please remove all version constraints from the import-packages

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,24 @@
+ org.eclipse.smarthome.core.types,
+ org.osgi.framework,
+ org.slf4j
+Service-Component: OSGI-INF/*
```
Please change to OSGI-INF/*.xml

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/DSCAlarmBindingConstants.java:
```diff
@@ -0,0 +1,110 @@
+	
+    public final static String PARTITION_STATUS = "partition_status";
+    public final static String PARTITION_ARM_MODE = "partition_arm_mode";
+	public final static String PARTITION_ARMED = "partition_armed";
```
The files seem to have a mix of tabs and spaces - please run the automatic formatter over all code, so that all tabs are replaced.

RSStephens commented:
I will review your comments and make changes accordingly.  Thanks Kai.

kaikreuzer commented:
Thanks - so far I only reviewed the meta-data, not yet the Java code. So more comments to come in the next days :-)

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/internal/protocol/APICommand.java:
```diff
@@ -0,0 +1,86 @@
+/**
+ * Class to create a API Command string.
+ *  
+ * @author Donn Renk - (https://code.google.com/p/homewatcher/)
```
WARNING: This project is licensed under GPL, which is incompatible with the EPL, i.e. we MUST NOT include any such code in openHAB. Please either:
- rewrite the code yourself
- get approval by the original author that he is ok to contribute this code under EPL to openHAB. Then add an Also-By footer to your commit record (see https://wiki.eclipse.org/Development_Resources/Contributing_via_Git#The_Commit_Record)

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/internal/protocol/APIMessage.java:
```diff
@@ -0,0 +1,840 @@
+ * 
+ * @author Donn Renk - (https://code.google.com/p/homewatcher/)
+ * @author Russell Stephens
+ * @since 1.6.0
```
please remove the since tag

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/internal/discovery/IT100BridgeDiscovery.java:
```diff
@@ -0,0 +1,54 @@
+
+	    while (ports.hasMoreElements()) {
+	        CommPortIdentifier port = (CommPortIdentifier) ports.nextElement();
+        	System.out.println("discoverBridge(): Port '" + port.getName() + "' Found!");
```
please remove the System.out statement

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/internal/discovery/IT100BridgeDiscovery.java:
```diff
@@ -0,0 +1,54 @@
+	        CommPortIdentifier port = (CommPortIdentifier) ports.nextElement();
+        	System.out.println("discoverBridge(): Port '" + port.getName() + "' Found!");
+	        if (port.getPortType() == CommPortIdentifier.PORT_SERIAL) {
+	        	System.out.println("discoverBridge(): Serial Port '" + port.getName() + "' Found!");
```
please remove the System.out statement

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/internal/discovery/IT100BridgeDiscovery.java:
```diff
@@ -0,0 +1,54 @@
+	        if (port.getPortType() == CommPortIdentifier.PORT_SERIAL) {
+	        	System.out.println("discoverBridge(): Serial Port '" + port.getName() + "' Found!");
+	        	logger.debug("discoverBridge(): Serial Port '{}' Found!",port.getName());
+	        	dscAlarmBridgeDiscovery.addIT100Bridge(port.getName());
```
you are taking ANY existing serial port here and adding it as a discovery result - this is not ok. You should only create a discovery result in case you have identified the IT-100 on the port.

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/internal/discovery/DSCAlarmBridgeDiscovery.java:
```diff
@@ -0,0 +1,166 @@
+		}else {
+			bridgeID = port.toUpperCase();			
+		}
+		
```
You also need to remove any invalid characters from the ID. I e.g. got the exception:
```
2015-10-06 21:31:08 [DEBUG] [.d.i.d.DSCAlarmBridgeDiscovery:163  ] - addBridge(): Error: {}java.lang.IllegalArgumentException: UID segment 'TTY.BLUETOOTH-INCOMING-PORT' contains invalid characters. Each segment of the UID must match the pattern [A-Za-z0-9_-]*.
	at org.eclipse.smarthome.core.thing.UID.validateSegment(UID.java:92)
	at org.eclipse.smarthome.core.thing.UID.<init>(UID.java:70)
	at org.eclipse.smarthome.core.thing.ThingUID.<init>(ThingUID.java:38)
	at org.openhab.binding.dscalarm.internal.discovery.DSCAlarmBridgeDiscovery.addIT100Bridge(DSCAlarmBridgeDiscovery.java:152)
	at org.openhab.binding.dscalarm.internal.discovery.IT100BridgeDiscovery.discoverBridge(IT100BridgeDiscovery.java:50)
```

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/internal/discovery/DSCAlarmBridgeDiscovery.java:
```diff
@@ -0,0 +1,166 @@
+						.withLabel("DSC IT-100 Bridge - " + port).build();
+				thingDiscovered(result);
+				
+				logger.debug("addBridge(): '{}' was added to Smarthome inbox.",result.getThingUID());
```
those things are logged by the framework already, so either remove it or reduce to TRACE.

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/internal/discovery/DSCAlarmBridgeDiscovery.java:
```diff
@@ -0,0 +1,166 @@
+						.withLabel("EyezOn Envisalink Bridge - " + ipAddress).build();
+				thingDiscovered(result);
+				
+				logger.debug("addBridge(): '{}' was added to Smarthome inbox.",result.getThingUID());
```
those things are logged by the framework already, so either remove it or reduce to TRACE.

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/internal/discovery/DSCAlarmBridgeDiscovery.java:
```diff
@@ -0,0 +1,166 @@
+	 * @param ipAddress
+	 */
+	public void addEnvisalinkBridge(String ipAddress) {
+		logger.debug("addBridge(): Adding new Envisalink Bridge on {} to Smarthome inbox",ipAddress);
```
those things are logged by the framework already, so either remove it or reduce to TRACE.

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/internal/discovery/EnvisalinkBridgeDiscovery.java:
```diff
@@ -0,0 +1,185 @@
+			networkInterface = NetworkInterface.getByInetAddress(localHost);
+		}
+		catch (UnknownHostException e) {
+			e.printStackTrace();
```
replace by a logging statement

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/internal/discovery/EnvisalinkBridgeDiscovery.java:
```diff
@@ -0,0 +1,185 @@
+			e.printStackTrace();
+		}
+		catch(SocketException e) {
+			e.printStackTrace();
```
replace by a logging statement

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/internal/discovery/EnvisalinkBridgeDiscovery.java:
```diff
@@ -0,0 +1,185 @@
+				try {
+					socket.close();
+				} catch (IOException e) {
+					e.printStackTrace();
```
replace by a logging statement

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/internal/discovery/EnvisalinkBridgeDiscovery.java:
```diff
@@ -0,0 +1,185 @@
+		logger.debug("   Last IP:          {}",longToIP(lastIP));
+
+		for(long ip=firstIP; ip <= lastIP; ip++) {
+			try {
```
you could do a try-with-resources here

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/internal/discovery/EnvisalinkBridgeDiscovery.java:
```diff
@@ -0,0 +1,185 @@
+		
+		setupIP();
+		
+		logger.debug("   Local IP Address: {} - {}",localHost.getHostAddress(), ipToLong(localHost.getHostAddress()));
```
probably also better trace level

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/internal/discovery/EnvisalinkBridgeDiscovery.java:
```diff
@@ -0,0 +1,185 @@
+		logger.debug("   First IP:         {}",longToIP(firstIP));
+		logger.debug("   Last IP:          {}",longToIP(lastIP));
+
+		for(long ip=firstIP; ip <= lastIP; ip++) {
```
I am not sure how this logic is exactly supposed to work.
In my case, my openHAB server has the IP 192.168.0.145 - the first IP is then 192.168.0.144 and the last IP is 192.168.0.146. What sense does this make...? Maybe the problem is that is also claims that my netmask is 255.255.255.255 (instead of 255.255.255.0)?

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/internal/discovery/EnvisalinkBridgeDiscovery.java:
```diff
@@ -0,0 +1,185 @@
+		for(long ip=firstIP; ip <= lastIP; ip++) {
+			try {
+				ipAddress = longToIP(ip);
+				if(!ipAddress.equals("10.1.32.6")) {
```
What is so special about this IP?

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/internal/discovery/EnvisalinkBridgeDiscovery.java:
```diff
@@ -0,0 +1,185 @@
+
+	private DSCAlarmBridgeDiscovery dscAlarmBridgeDiscovery = null;
+	private String ipAddress;
+	private int envisalinkBridgePort = 4025;
```
this could be a constant, right?

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/internal/discovery/EnvisalinkBridgeDiscovery.java:
```diff
@@ -0,0 +1,185 @@
+	private DSCAlarmBridgeDiscovery dscAlarmBridgeDiscovery = null;
+	private String ipAddress;
+	private int envisalinkBridgePort = 4025;
+	private int connectionTimeout = 5;
```
this could be a constant, right?

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/internal/discovery/EnvisalinkBridgeDiscovery.java:
```diff
@@ -0,0 +1,185 @@
+				socket.connect(new InetSocketAddress(ipAddress, envisalinkBridgePort), connectionTimeout);
+				
+				if (socket.isConnected()) {
+					System.out.printf("Bridge Found! - %s\n", ipAddress);
```
please remove the System.out

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/internal/discovery/EnvisalinkBridgeDiscovery.java:
```diff
@@ -0,0 +1,185 @@
+				socket.setReceiveBufferSize(32);
+				socket.connect(new InetSocketAddress(ipAddress, envisalinkBridgePort), connectionTimeout);
+				
+				if (socket.isConnected()) {
```
Hm, is there again a way to make sure that it is the correct device?
The port 4025 is actually reserved for partimage by IANA: http://www.speedguide.net/port.php?port=4025, so there is a certain probability that you discover the wrong server.

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/internal/discovery/EnvisalinkBridgeDiscovery.java:
```diff
@@ -0,0 +1,185 @@
+		
+		ipArray = ip.split("\\.");
+        if (ipArray.length != 4) {
+            throw new NumberFormatException("Invalid IP address: " + ip);
```
You aren't catching this exception anywhere.

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/internal/discovery/EnvisalinkBridgeDiscovery.java:
```diff
@@ -0,0 +1,185 @@
+        for (int n = 0; n < ipArray.length; n++) {
+            long value = Long.parseLong(ipArray[n]);
+            if (value != (value & 0xff)) {
+                throw new NumberFormatException("Invalid IP address: " + ip);
```
You aren't catching this exception anywhere.

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/handler/ZoneThingHandler.java:
```diff
@@ -0,0 +1,254 @@
+	 * {@inheritDoc}
+	 */
+	@Override
+	public void dscAlarmEventRecieved(EventObject event, Thing thing) {
```
typo: dscAlarmEventReceived

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/handler/ZoneThingHandler.java:
```diff
@@ -0,0 +1,254 @@
+	public void dscAlarmEventRecieved(EventObject event, Thing thing) {
+
+		if(thing != null) {
+			if(getThing() == thing) {
```
better to a equals(thing)

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/handler/DSCAlarmBaseThingHandler.java:
```diff
@@ -0,0 +1,350 @@
+		logger.debug("Initializing DSC Alarm Thing handler - Thing Type: {}; Thing ID: {}.", dscAlarmThingType, this.getThing().getUID());
+
+		getConfiguration(dscAlarmThingType);				
+		updateStatus(ThingStatus.OFFLINE);
```
Why offline? If you do not know whether it is online or offline, better don't set any status here.

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/handler/DSCAlarmBaseThingHandler.java:
```diff
@@ -0,0 +1,350 @@
+	public void refreshThing() {
+
+		if(getDSCAlarmBridgeHandler() != null) {
+			updateStatus(dscAlarmBridgeHandler.getThing().getStatus());
```
you can imho remove this. You get calls on bridge connect and disconnect and therefore you can set the status there, which is enough.

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/handler/DSCAlarmBaseThingHandler.java:
```diff
@@ -0,0 +1,350 @@
+		
+		if(bridge.getThing().getUID().equals(getThing().getBridgeUID())) {
+			this.setThingRefreshed(false);
+			updateStatus(ThingStatus.OFFLINE);
```
remove this, as it is set by the framework (and then even with the status detail BRIDGE_OFFLINE)

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/src/main/java/org/openhab/binding/dscalarm/handler/DSCAlarmBaseBridgeHandler.java:
```diff
@@ -0,0 +1,693 @@
+	 * Check if things have changed.
+	 */
+	public void checkThings() {
+		logger.debug("Checking Things!");
```
rather trace

kaikreuzer commented:
@RSStephens Thanks, I am through with my review.
Please apply the code formatter on the code.
The biggest issue is currently the included GPL code - I hope you find an easy solution for this!

RSStephens commented:
Hello Kai,

I saw all the emails.  It looks like I have my work cut out for me.  I think I will look at rewriting the GPL code, I do have consent to use it from the author but that was back on openhab 1, and most of it has been modified for use in openhab.  So I will see what I can do with it.  Thanks for your review I will start making the changes.  I appreciate it.

buildhive commented:
[openHAB » openhab2 #34](https://buildhive.cloudbees.com/job/openhab/job/openhab2/34/) FAILURE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[openHAB » openhab2 #45](https://buildhive.cloudbees.com/job/openhab/job/openhab2/45/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[openHAB » openhab2 #46](https://buildhive.cloudbees.com/job/openhab/job/openhab2/46/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[openHAB » openhab2 #47](https://buildhive.cloudbees.com/job/openhab/job/openhab2/47/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

kaikreuzer commented:
@RSStephens Please ping me whenever you think you are done and I should do a final review round. Thanks!

buildhive commented:
[openHAB » openhab2 #63](https://buildhive.cloudbees.com/job/openhab/job/openhab2/63/) FAILURE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[openHAB » openhab2 #64](https://buildhive.cloudbees.com/job/openhab/job/openhab2/64/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

RSStephens commented:
Hello Kai,

I believe it is ready for a final review.  Here is a list of some of the changes made since your original review:

* Removed 'bridge' from name of DSC Alarm bridges.
* Changed parameter label from 'Envisalink Bridge IP' to 'IP Address'.
* Added '<context>password</context>' to parameter.
* Changed parameter description from 'The Poll Period.' to 'The Poll Period (minutes).'
* Used <options> in some channel-types instead of max, min.
* Consolidated channel-types for use with channels of the same type.
* Removed conversion constraints in MANIFEST.MF file.
* Changed 'OSGI-INF/*' to 'OSGI-INF/*.xml' in MANIFEST.MF file.
* Removed GPL code.
* Removed 'System.out' statements
* Rewrote the serial (IT100) bridge discovery class to prompt for a specific response for verification of the bridge.
* Rewrote the TCP (Envisalink) bridge discovery class to wait for a specific response for verification of the bridge.
* Fixed serial bridge discovery to remove invalid characters in the bridge ID.
* Changed DEBUG log statements to TRACE.
* Replaced Exception print statements with logging statements.
* Replaced 'try-catch-finally' with 'try-with-resources in TCP bridge discovery.
* Converted some variables to constants in the bridge discovery classes.
* Catching thrown exception now.
* Removed updating thing status to OFFLINE on initialization.
* Ran automatic formatting.

Thank you!

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/ESH-INF/thing/envisalinkbridge.xml:
```diff
@@ -0,0 +1,64 @@
+	xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 org.eclipse.smarthome.thing-description.xsd">
+
+	<bridge-type id="envisalink">
+		<label>EyezOn Envisalink Bridge</label>
```
I'd remove the word "Bridge" here

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/ESH-INF/thing/envisalinkbridge.xml:
```diff
@@ -0,0 +1,64 @@
+
+		<channels>
+			<channel id="bridge_reset" typeId="reset">
+				<label>Reset Envisalink Bridge</label>
```
I'd remove the word "Bridge" here

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/ESH-INF/thing/it100bridge.xml:
```diff
@@ -0,0 +1,48 @@
+	xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 org.eclipse.smarthome.thing-description.xsd">
+
+	<bridge-type id="it100">
+		<label>DSC IT-100 Bridge</label>
```
I'd remove the word "Bridge" here

kaikreuzer commented on addons/binding/org.openhab.binding.dscalarm/ESH-INF/thing/it100bridge.xml:
```diff
@@ -0,0 +1,48 @@
+
+		<channels>
+			<channel id="bridge_reset" typeId="reset">
+				<label>Reset IT100 Bridge</label>
```
I'd remove the word "Bridge" here

kaikreuzer commented:
Hi @RSStephens , thanks a lot for the update! I had a glance over it and it looks all good. Does the list of changes that you did cover all my comments or was there anything where you think I was wrong or what might need any discussion?
One last question: Why did you change the start*.* scripts? Is that through the auto-formatting or did you actually change the content? If not, please revert these files again.

Once you are done, please squash your commits into one and we should be ready to merge. Thanks!

RSStephens commented:
Hello Kai, the list is based on your comments and I don't believe there was anything that I had any issues with.  I don't know about the start files.  I didn't do any edits in them.  Maybe it was during the auto-formatting.  I looked at the commits and the last one was on October 16, which came over when I did an upstream merge, so it must have happened in the auto-formatting.  It doesn't let me revert the files either.

I will squash my commits.  Thanks.

kaikreuzer commented:
You should actually never do an upstream merge into your branch, but rather rebase your branch on the latest master. This avoids having changed from the master within your PR. After squashing your commits, you could try to simply revert the changes to the start files and amend your commit - this should make these diffs disappear.

buildhive commented:
[openHAB » openhab2 #65](https://buildhive.cloudbees.com/job/openhab/job/openhab2/65/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[openHAB » openhab2 #66](https://buildhive.cloudbees.com/job/openhab/job/openhab2/66/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[openHAB » openhab2 #67](https://buildhive.cloudbees.com/job/openhab/job/openhab2/67/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

kaikreuzer commented:
You do not seem to be too successful with the squashing yet, right?

RSStephens commented:
No, not at all. Sorry.  Any ideas?

kaikreuzer commented:
You can try some git magic (make sure you have a backup of your local repo ;-)):
From your branch, do a
```
git reset --soft origin/master
git reset distribution/openhabhome/*.bat
git commit
git push -f https://github.com/RSStephens/openhab2.git DSCAlarm
```

RSStephens commented:
Wow! That is magic.  Thanks, Kai.

buildhive commented:
[openHAB » openhab2 #68](https://buildhive.cloudbees.com/job/openhab/job/openhab2/68/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[openHAB » openhab2 #69](https://buildhive.cloudbees.com/job/openhab/job/openhab2/69/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

kaikreuzer commented:
But we are still left with 3 commits... I would be able to simply cherry pick https://github.com/RSStephens/openhab2/commit/ff2e9ae07262349cbe69d21ba87c3410ae18e386, but I noticed that you have missed signing it off. Could you please amend your commit and add the sign-off footer to your commit message?

buildhive commented:
[openHAB » openhab2 #72](https://buildhive.cloudbees.com/job/openhab/job/openhab2/72/) FAILURE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

kaikreuzer commented:
You are getting closer. Now only fix your commit message to only include your change and not claim it is signed off by Marcel Verpaalen.
And the PR says it has merge conflicts, so you might need to rebase your branch again.

buildhive commented:
[openHAB » openhab2 #73](https://buildhive.cloudbees.com/job/openhab/job/openhab2/73/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

RSStephens commented:
Hello Kai, I believe I have it now.  This has been quite an exercise in the use of Git.  I appreciate all your help.  Thank you.

kaikreuzer commented:
Congratulations :-)
Now you can call yourself a git guru.
